### PR TITLE
Add type hints and initial Mypy configuration

### PR DESCRIPTION
**Pull request by THS-on**
This PR adds full type hints to:
- `ima_ast.py`
- `failure.py`
- `algorithms.py`
- `keylime_loggin.py`
- `json.py`

I also added a Mypy configuration with `strict` set to true which should enable most of the Mypy checks.

Once more parts of Keylime have type hints I will add this to the tox configuration.

Part of #929

### CODE REVIEW

File: keylime/failure.py [hunk 1]
```diff
@@ -92,18 +95,18 @@ class Failure:
     events: List[Event]
     recoverable: bool
     highest_severity: Optional[SeverityLabel]
-    _component: Optional[Component]
+    _component: Component
     _sub_components: Optional[List[str]]
 
-    def __init__(self, component, sub_components=None):
+    def __init__(self, component: Component, sub_components: Optional[List[str]] = None):
```
Comment (mpeters) [inline keylime/failure.py:101]: Just a question on style here: Does the type hints syntax prefer whitespace around the equals sign?

Comment (THS-on) [reply keylime/failure.py:101]: I think so: https://peps.python.org/pep-3107/#parameters.

**Comment by mpeters** [comment]
Overall I like this change as it should prevent some accidental bugs in the future. Being unfamiliar with Mypy, are there any performance considerations?

**Comment by THS-on** [comment]
>  Being unfamiliar with Mypy, are there any performance considerations?

Type hints are ignored at runtime, so there should be no performance impact other than the hints are getting parsed and then ignored.

**Review by maugustosilva — APPROVED** [review]

**Comment by THS-on** [comment]
@kkaarreell why did this e2e CI pass?

**Comment by kkaarreell** [comment]
There is some issue with tmt plan on c9s and c8s where prepare steps are replaced, not added. As a result c8s and c9s are run against upstream master. I have just found out few minutes back.

**Pull request merged**